### ISSUE TITLE
Missing dr key causes error in FetchData

### ISSUE THREAD

**Issue by lazappi** [satijalab/seurat#1089]
If a (v2.3.4) seurat object has something in the dr slot which does not have a key set this causes an obscure error in `FetchData` causing other functions to fail in a non-obvious way. Example using the `pbmc_small` data:

```r
head(FetchData(pbmc_small, "nGene"))
#                nGene
# ATGCCAGAACGACT    47
# CATGGCCTGTGCAT    52
# GAACCTGATGAACC    50
# TGACTGGATTCTCA    56
# AGTCAGACTGCACA    53
# TCTGATACACGTGT    48

names(pbmc_small@dr) [1] "pca"  "tsne"
# [1] "pca"  "tsne"

pbmc_small@dr$pca_broken <- pbmc_small@dr$pca
pbmc_small@dr$pca_broken@key <- character(0)

head(FetchData(pbmc_small, "nGene"))
# Error in parse(text = paste0("data.use <- object@dr$", names(x = var.options[which(i ==  : 
#   <text>:1:23: unexpected '@'
# 1: data.use <- object@dr$@
                          ^ 
```

Seems reasonable for missing keys to cause problems but from the error it's not at all clear this is the problem, maybe that can be checked somewhere and a better error thrown? This happened after converting from a `SingleCellExperiment` object so perhaps how this part of the conversion is done could be looked at as well?

**Comment by mojaveazure** [comment]
This should be fixed in Seurat v3; Seurat enforces keys being set and valid when using the proper setters and accessors. Since `SingleCellExperiment` &rarr; `Seurat` conversion is present in v3 (`as.Seurat`), could you let us know if something failing with that?

**Comment by lazappi** [comment]
I don't have Seurat v3 installed anywhere at the moment but looking at the code I think there might still be the same problem. In my case the the `reducedDim` in my `SingleCellExperiment` object had no column names which meant during the conversion (here https://github.com/satijalab/seurat/blob/develop/R/conversion.R#L320) the key was set to be an empty string. The object still converted fine but there was an error when that key was attempted to be used. The simplest solution is probably just to check the key during conversion and raise an error if it is empty.

**Comment by andrewwbutler** [comment]
Hi Luke,

Thanks for reporting! We do check for this better in V3 (on the `release/3.0` branch) and it will actually error out in the conversion function if the `reducedDim` doesn't have column names with a slightly more informative message. 

However, I think we can improve the conversion function here by just setting the key to the name of the `reducedDim` if it can't guess the key from the column names. This should work as of ca39dda

**Comment by lazappi** [comment]
Sounds like a reasonable fix to me 👍

**Comment by carmensandoval** [comment]
Hi all,

Previously, in v2, I was able to solve this problem by setting the `@key` slot of `seurat.object@dr$custom.dr`, such as:

`seurat.object@dr$fastPCA@key <- "fastPCA_"`

After I updated to v3, I'm no longer able to fix this. I get this error message with the above command: 

`Error in getClass(cl) : "dim.reduction" is not a defined class`

Because of this error, I'm not able to update my v2 object to v3. 

`seurat.object.v3 <- UpdateSeuratObject(seurat.object)`

fails with the error: 

```
Error in if (grepl(pattern = "^[[:alnum:]]+_", x = key)) { :
  argument is of length zero
```

Are there any workarounds to this? I don't have Seurat V2 anymore to fix this `@key` slot and update it.
